### PR TITLE
To allow ts-proto to create a NestJS interfaces for use on both client / server

### PR DESCRIPTION
**Pull request by iangregsondev**
Still left to do is update the Readme - once the API has stabilised.

To restart a dialogue about changes.

Currently, the changes ONLY affect the interface. Need to look at tests if we decide any of these changes propagate to any modified concrete types.

### CODE REVIEW

**Comment by iangregsondev** [comment]
@stephenh apologies before about the bot closing the previous pull request. I have removed the bot and created this pull request.

Thank you.

**Comment by iangregsondev** [comment]
Thanks, I am starting to push changes now - little by little.

File: src/main.ts [hunk 1]
```diff
@@ -58,6 +58,9 @@ export type Options = {
   outputEncodeMethods: boolean;
   outputJsonMethods: boolean;
   outputClientImpl: boolean;
+  addGrpcMetadata: boolean;
```
Comment (iangregsondev) [inline src/main.ts:61]: I have added a nestJS which encapsulate the interface only options that existed previously, it also sets the method names to "lowercase". This is needed by nestjs.

I can't encapsulate the "addGrpcmetaData" and "returnObservable" because this depends on the user.

File: src/main.ts [hunk 2]
```diff
@@ -967,6 +970,11 @@ function generateService(
 
   let index = 0;
   for (const methodDesc of serviceDesc.method) {
+
+    if (options.nestJs) {
```
Comment (iangregsondev) [inline src/main.ts:974]: This is where I change the case of the methods to lowercase

File: src/utils.ts [hunk 3]
```diff
@@ -67,6 +70,22 @@ export function optionsFromParameter(parameter: string): Options {
     if (parameter.includes('outputClientImpl=false')) {
       options.outputClientImpl = false;
     }
+
+    if (parameter.includes('nestJs=true')) {
```
Comment (iangregsondev) [inline src/utils.ts:74]: Notice that the "addGrpcMetadata" and "returnObservable" are ONLY available if its a nestjs run. It really doesn't make a lot of sense for anything but nest.

Comment (iangregsondev) [reply src/utils.ts:74]: This also keeps things nice and backward compatible.

File: README.markdown [hunk 4]
```diff
@@ -150,6 +150,15 @@ Supported options:
 
 (I.e. you want only interface declarations for your Protobuf types, then pass all three of `outputEncodeMethods`, `outputJsonMethods`, and `outputClientImpl` as `false`, i.e. `--ts_proto_opt=outputEncodeMethods=false,outputJsonMethods=false,outputClientImpl=false`.)
 
+The following options are very useful for NestJS and available if `nestJs` is set as `true`
```
Comment (iangregsondev) [inline README.markdown:153]: I have tried to give a little bit of insight on what the options do. I don't think we need to give specific examples here of what to do with the final interfaces, everyone coming from nestJS will know.

What I will do is add some context about what metadata is.

File: README.markdown [hunk 5]
```diff
@@ -150,6 +150,18 @@ Supported options:
...
+
+(Setting `nestJs` as `true` will implicitly set `outputEncodeMethods`, `outputJsonMethods`, and `outputClientImpl` as `false`)
+
+* If you pass `--ts_proto_opt=addGrpcMetadata=true`, the last argument accepts the grpc metadata, this is useful to send additional information with the call.
+* If you pass `--ts_proto_opt=returnObservable=true`, the return type will be of type Observable<T> rather than Promise<T>
+
+The NestJS option will output an interface declaration that is specifically compatible with NestJS and can be used both on the client and server.
+
+The grpc metadata parameter is generated as an optional and is really useful for passing down extra information (meta data) with the call (from the client ot the server). There are many use cases for generating the metadata parameter, for example, to send down a JWT Access Token.
```
Comment (iangregsondev) [inline README.markdown:162]: I have tried to give some insight here about the metadata.

**Comment by iangregsondev** [comment]
@stephenh I am going to have a look at implementing some tests for the new functionality.

I have pushed my latest changes and added some conversations above. Please feel free to come back and let me know your faults.

I might ask for your help with the new tests soon

**Comment by iangregsondev** [comment]
> in case it's helpful, i've ran the integration tests on my machine and am getting the same errors even on https://github.com/stephenh/ts-proto/tree/master , so that failure doesn't seem to be related to these changes:
> 
> `
> FAIL integration/simple/simple-test.ts
> ● Test suite failed to run
> 
> ```
> integration/simple/simple-test.ts:74:12 - error TS2790: The operand of a 'delete' operator must be optional.
> 
> 74     delete s1.blobs;
>               ~~~~~~~~
> integration/simple/simple-test.ts:160:12 - error TS2790: The operand of a 'delete' operator must be optional.
> 
> 160     delete s1.blobs;
> ```
> 
> `

Hi - I managed to fix the 2 errors. I think that a new version of typescript doesn't allow you to delete a property / field that is NOT optional.

2 fixes here,

1. ts-ignore that line - these are only tests :-)

2. Change the field "blob" to an optional - this fixes that issue but that you need to check for "undefined" in other parts of the code as it fails now with the typical "blob could be undefined" etc

@stephenh need your input here.

@tlinford I got a lot more errors after this - specifically in the 

```
 FAIL  integration/simple/numbers-test.ts
```

with some errors like

```
 FAIL  integration/simple/numbers-test.ts
  ● Test suite failed to run

    Could not find source file: '/Users/ian/Development/Personal/typescript/Forks/ts-proto/integration/simple/numbers-test.ts'.

      at getValidSourceFile (node_modules/typescript/lib/typescript.js:138609:29)
      at Object.getSemanticDiagnostics (node_modules/typescript/lib/typescript.js:138832:36)
      at doTypeChecking (node_modules/ts-jest/dist/compiler/language-service.js:11:35)
      at node_modules/ts-jest/dist/compiler/language-service.js:96:25
          at Array.forEach (<anonymous>)
      at compileFn (node_modules/ts-jest/dist/compiler/language-service.js:94:26)
      at Object.compile (node_modules/ts-jest/dist/compiler/instance.js:90:25)
      at TsJestTransformer.process (node_modules/ts-jest/dist/ts-jest-transformer.js:94:41)
      at ScriptTransformer.transformSource (node_modules/@jest/transform/build/ScriptTransformer.js:446:35)
      at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:525:40)
```

there are quite a few - do you get these after fixing the above issue ?

I notice the version of typescript that is being used is a beta 

```
    "typescript": "^3.9.0-beta"
```

This is really new :-) I would suggest not using that. 

I don't know if you use renovate or similar to update the package versions ? Probably should not accept a typescript beta.

**Comment by iangregsondev** [comment]
> @tlinford I got a lot more errors after this - specifically in the
> 
> ```
>  FAIL  integration/simple/numbers-test.ts
> ```
> 
> with some errors like
> 
> ```
>  FAIL  integration/simple/numbers-test.ts
>   ● Test suite failed to run
> 
>     Could not find source file: '/Users/ian/Development/Personal/typescript/Forks/ts-proto/integration/simple/numbers-test.ts'.
> 
>       at getValidSourceFile (node_modules/typescript/lib/typescript.js:138609:29)
>       at Object.getSemanticDiagnostics (node_modules/typescript/lib/typescript.js:138832:36)
>       at doTypeChecking (node_modules/ts-jest/dist/compiler/language-service.js:11:35)
>       at node_modules/ts-jest/dist/compiler/language-service.js:96:25
>           at Array.forEach (<anonymous>)
>       at compileFn (node_modules/ts-jest/dist/compiler/language-service.js:94:26)
>       at Object.compile (node_modules/ts-jest/dist/compiler/instance.js:90:25)
>       at TsJestTransformer.process (node_modules/ts-jest/dist/ts-jest-transformer.js:94:41)
>       at ScriptTransformer.transformSource (node_modules/@jest/transform/build/ScriptTransformer.js:446:35)
>       at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:525:40)
> ```

Ok I needed to regenerate the test files :-) 

Seems to be passing now - but

The 2 additional 

```
    // @ts-ignore
```

needed to be left there.

File: README.markdown (hunk 4 above)
Comment (stephenh) [reply README.markdown:153]: Nice, thanks! I word-smithed this a bit, please check master and let me know if something doesn't sound right.

File: src/main.ts (hunk 2 above)
Comment (stephenh) [reply src/main.ts:974]: Nice, I ended up adding a separate `lowerCaseServiceNames` flag that `nestJs` changes to `true`.

File: src/main.ts [hunk 6]
```diff
@@ -975,7 +983,19 @@ function generateService(
     maybeAddComment(info, text => (requestFn = requestFn.addJavadoc(text)));
 
     requestFn = requestFn.addParameter('request', requestType(typeMap, methodDesc));
-    requestFn = requestFn.returns(responsePromise(typeMap, methodDesc));
+
+    // Use metadata as last argument for interface only configuration
+    if (!options.outputClientImpl && !options.outputEncodeMethods && !options.outputJsonMethods && options.addGrpcMetadata) {
```
Comment (stephenh) [inline src/main.ts:988]: I'm pretty sure it's fine, but I simplified this to `if (options.addGrpcMetadata)` when I merged to master...

Hm, in retrospect, `outputEncodeMethods` and `outputJsonMethods` really shouldn't matter for adding the grpc metadata, although `outputClientImpl` would probably break b/c it'll be outputting impls that don't know about the metadata.

I think I'll keep this simplification as is and then push that out (eventually/maybe) to a config option check that addGrpcMetadata and outoutClientImpl are not compatible.

File: src/main.ts [hunk 7]
```diff
@@ -975,7 +983,19 @@ function generateService(
...
+      requestFn = requestFn.addParameter('metadata?', "Metadata@grpc");
+    }
+
+    // Return observable for interface only configuration and passing returnObservable=true
+    if (!options.outputClientImpl && !options.outputEncodeMethods && !options.outputJsonMethods && options.returnObservable) {
```
Comment (stephenh) [inline src/main.ts:993]: Simplified with same rationale, but in retrospect caveat about outputclientimpl, as above.

**Comment by stephenh** [comment]
Closing per comment that it's already merged to master.

File: integration/nestjs-metadata-observables/parameters.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+nestJs=true,returnObservable=true,addGrpcMetadata=true
```
[the remark this answers is not in the document]
Comment (iangregsondev) [reply integration/nestjs-metadata-observables/parameters.txt:1]: I think not, as not everyone wants to use observables, although they should :-) NestJS is built upon observables :-)

And addGrpcMetadata - i would say NO, this is kind of a edge case in my honest opinion - although they are optional - so it wouldn't hurt.

File: integration/simple/simple-test.ts [hunk 9]
```diff
@@ -65,6 +65,8 @@ describe('simple', () => {
       blobs: []
     };
     const s2 = PbSimple.toObject(PbSimple.decode(Simple.encode(s1).finish()));
+
+    // @ts-ignore
```
[the remark this answers is not in the document]
Comment (iangregsondev) [reply integration/simple/simple-test.ts:69]: It was failing, maybe i was using a later version of typescript - not sure.

If it is working now, then great.

File: src/main.ts (hunk 2 above)
Comment (iangregsondev) [reply src/main.ts:974]: Great!
